packages/solana-swap-provider/lib/SolanaSwapProvider.ts

```diff
@@ -0,0 +1,282 @@
+      buyerAccount,
+      createClaimBuffer(secret),
+      programId
+    )
+
+    return await this.getMethod('sendTransaction')({
+      instructions: [transactionInstruction]
+    })
+  }
+
+  async refundSwap(swapParams: SwapParams, initiationTxHash: string, fee: number): Promise<Transaction<any>> {
+    await this.verifyInitiateSwapTransaction(swapParams, initiationTxHash)
+
+    const [initTransaction] = await this.getMethod('getParsedAndConfirmedTransactions')([initiationTxHash])
+
+    const { programId, seller } = initTransaction._raw
+
+    const [programAccount] = await this.getMethod('getProgramAccounts')(programId.toString())
+
+    if (!programAccount) {
+      throw new Error('AccountDoesNotExist')
+    }
+
+    const appAccount = new PublicKey(programAccount.pubkey)
+    const sellerAccount = new PublicKey(seller)
+
+    const transactionInstruction = await this._collectLamports(
+      appAccount,
+      sellerAccount,
+      createRefundBuffer(),
+      programId.toString()
+    )
+
+    return await this.getMethod('sendTransaction')({
+      instructions: [transactionInstruction]
+    })
+  }
+
+  async verifyInitiateSwapTransaction(swapParams: SwapParams, initiationTxHash: string): Promise<boolean> {
+    this._validateSwapParams(swapParams)
+
+    const [initTransaction] = await this.getMethod('getParsedAndConfirmedTransactions')([initiationTxHash])
+
+    return this._compareParams(swapParams, initTransaction._raw)
+  }
+
+  async _collectLamports(
+    appAccount: any,
+    recipient: PublicKey,
+    data: any,
+    _programId: string
+  ): Promise<TransactionInstruction> {
+    const appAccountPubkey = appAccount.publicKey || appAccount
+    const signer = await this.getMethod('getSigner')()
+    const programId = new PublicKey(_programId)
+
+    return new TransactionInstruction({
+      keys: [
+        { pubkey: signer.publicKey, isSigner: true, isWritable: true },
+        { pubkey: appAccountPubkey, isSigner: false, isWritable: true },
+        { pubkey: recipient, isSigner: false, isWritable: true }
+      ],
+      programId,
+      data
+    })
+  }
+
+  async initTxParams(programId: string): Promise<any> {
+    const accounts = await this.getMethod('getProgramAccounts')(programId)
+
+    const accountInfo = await this.getMethod('getAccountInfo')(accounts[0].pubkey.toString())
+
+    return this._deserialize(accountInfo.data)
+  }
+
+  async fundSwap(): Promise<null> {
+    return null
+  }
+
+  _validateSwapParams(swapParams: SwapParams): void {
+    validateValue(swapParams.value)
+    validateSecretHash(swapParams.secretHash)
+    validateExpiration(swapParams.expiration)
+    this._validateAddress(swapParams.recipientAddress as string)
+    this._validateAddress(swapParams.refundAddress as string)
+  }
+
+  _validateAddress(address: string): boolean {
+    return typeof address === 'string' && address.length === 44
+  }
+
+  _compareParams(swapParams: SwapParams, transactionParams: InitData): boolean {
```

Comment:
You should use similar method to this one:

https://github.com/liquality/chainabstractionlayer/blob/dev/packages/near-swap-provider/lib/NearSwapProvider.ts#L110

Also, you have 2 `_compareParams`